Content.Shared/GameObjects/Components/Mobs/State/SharedMobStateComponent.cs

```diff
@@ -0,0 +1,336 @@
+            {
+                case DamageChangedMessage msg:
+                    if (msg.Damageable.Owner != Owner)
+                    {
+                        break;
+                    }
+
+                    UpdateState(msg.Damageable.TotalDamage);
+
+                    break;
+            }
+        }
+
+        public bool IsAlive()
+        {
+            return CurrentState?.IsAlive() ?? false;
+        }
+
+        public bool IsCritical()
+        {
+            return CurrentState?.IsCritical() ?? false;
+        }
+
+        public bool IsDead()
+        {
+            return CurrentState?.IsDead() ?? false;
+        }
+
+        public bool IsIncapacitated()
+        {
+            return CurrentState?.IsIncapacitated() ?? false;
+        }
+
+        public (IMobState state, int threshold)? GetState(int damage)
+        {
+            foreach (var (threshold, state) in _lowestToHighestStates.Reverse())
+            {
+                if (damage >= threshold)
+                {
+                    return (state, threshold);
+                }
+            }
+
+            return null;
+        }
+
+        public bool TryGetState(
+            int damage,
+            [NotNullWhen(true)] out IMobState? state,
+            out int threshold)
+        {
+            var highestState = GetState(damage);
+
+            if (highestState == null)
+            {
+                state = default;
+                threshold = default;
+                return false;
+            }
+
+            (state, threshold) = highestState.Value;
+            return true;
+        }
+
+        public (IMobState state, int threshold)? GetEarliestIncapacitatedState(int minimumDamage)
+        {
+            foreach (var (threshold, state) in _lowestToHighestStates)
+            {
+                if (!state.IsIncapacitated())
+                {
+                    continue;
+                }
+
+                if (threshold < minimumDamage)
+                {
+                    continue;
+                }
+
+                return (state, threshold);
+            }
+
+            return null;
+        }
+
+        public bool TryGetEarliestIncapacitatedState(
+            int minimumDamage,
+            [NotNullWhen(true)] out IMobState? state,
+            out int threshold)
+        {
+            var earliestState = GetEarliestIncapacitatedState(minimumDamage);
+
+            if (earliestState == null)
+            {
+                state = default;
+                threshold = default;
+                return false;
+            }
+
+            (state, threshold) = earliestState.Value;
+            return true;
+        }
+
+        private void RemoveState(bool syncing = false)
+        {
+            var old = CurrentState;
+            CurrentState = null;
+            CurrentThreshold = null;
+
+            UpdateState(old, null);
+
+            if (!syncing)
+            {
+                Dirty();
+            }
+        }
```

Comment:
When would syncing ever be used?

Reply:
HandleComponentState, I'm not sure if calling Dirty while syncing values with the server is a problem or not.

Reply:
NFI then, some other stuff calls Dirty() when using handlecomponentstate (transform and physics I think).